courses/gnatmetric/010_gnatmetric.rst

```diff
@@ -0,0 +1,352 @@
+
+**All Declarations**
+ Total number of objects declared
+
+**All Statements**
+ Total number of statements in file(s)
+
+**All Subprogram Bodies**
+ Total number of subprograms in file(s)
+
+**All Type Definitions**
+ Total number of types in file(s)
+
+**Logical SLOC**
+ Total of declarations plus statements
+
+**Public Subprograms**
+ Count of subprograms declared in visible part of package
+
+**Public Types**
+ Count of types (not subtypes) declared in the visible part of a package plus in the visible part of a generic nested package
+
+**Maximal Construct Nesting**
+ Maximal nesting level of composite syntactic constructs
+
+**Maximum Unit Nesting**
+ Maximal static nesting level of inner program units
+
+------------------------------
+Complexity Metrics Explained
+------------------------------
+
+**Average Complexity**
+    Total Cyclomatic Complexity divided by total number of subprograms
+
+**Cyclomatic Complexity**
+    McCabe cyclomatic complexity (number of independent paths in the control flow graph)
+
+**Essential Complexity**
+    McCabe essential complexity (cyclomatic complexity after removing blocks with single entry/exit points)
+
+**Expression Complexity**
+    Complexity introduced by short-circuit control forms only
+
+**Maximum Loop Nesting**
+    Maximum depth of nested loops
+
+**Statement Complexity**
+    Complexity introduced by control statements only, without taking into account short-circuit forms 
+
+---------------------------------
+Understanding McCabe Complexity
+---------------------------------
+
+http://www.mccabe.com/pdf/mccabe-nist235r.pdf
+
++ Given a control flow graph of a program
+
+  + E - number of edges
+  + N - number of nodes
+  + P - number of connected components (exit nodes)
+
++ The complexity is computed by:
+
+  + E - N + 2 * P
+
++ Aimed a measuring the complexity of execution paths
++ Needs to be adapted for each language
+
+----------------
+McCabe Example
+----------------
+
+.. columns::
+
+  .. column::
+
+   .. code:: Ada
+
+      if A then
+         Put_Line ("A");
+      else
+         Put_Line ("!A");
+      end if;
+
+      if B or else C then
+         Put_Line ("BC");
+      end if;
+
+  .. column::
+
+     .. image:: ../../images/gnatmetric/cyclomatic_complexity_edges_and_nodes.jpg
+
+     9 edges - 7 nodes + 2 * 1 exit = complexity 4
+
+------------------------------
+Coupling Metrics Explained
+------------------------------
+
++ Measures dependencies between given entity and other entities in the program
+
+  + High coupling may signal potential issues with maintainability
+
++ Metrics computed:
+
+  **Object-oriented coupling**
+     Classes in traditional object-oriented sense
+
+  **Unit coupling**
+     All units making up a program
+
+  **Control coupling**
+     Dependencies between unit and other units that contain subprograms
+
+------------------------------
+Coupling Metrics
+------------------------------
+
++ Uses Ada's approach to definition of *class*
```

Comment:
ARM definiton of class is

> A class of types is a set of types that is closed under derivation

That is, the case for most Ada type, e.g. `Integer` and its children, which are not included in remaining bullet points.
Did you mean "Based on Ada's class-wide polymorphism" ?

Reply:
Added indication that only polymorphic classes are measured